lightning-payjoin/src/lib.rs

```diff
@@ -1 +1,316 @@
+
+	fn validate_payjoin_request(
+		&self, body: &[u8], query: &str, headers: RequestHeaders, min_fee_rate: Option<FeeRate>,
+	) -> Result<ProvisionalProposal, Error> {
+		let proposal = payjoin::receive::UncheckedProposal::from_request(body, query, headers)?;
+		let proposal =
+			proposal.check_broadcast_suitability(min_fee_rate, |t| Ok(self.can_broadcast(t)?))?;
+		let proposal = proposal.check_inputs_not_owned(|script| {
+			Ok(self.check_inputs_not_owned(&script.to_owned())?)
+		})?;
+		let proposal = proposal.check_no_mixed_input_scripts()?;
+		let proposal = proposal.check_no_inputs_seen_before(|_outpoint| Ok(false))?;
+		let original_proposal = proposal.clone().identify_receiver_outputs(|script| {
+			Ok(self.identify_my_outputs(&script.to_owned())?)
+		})?;
+		Ok(original_proposal)
+	}
+
+	async fn get_next_channel(&self, payjoin_amount: u64) -> Option<scheduler::ScheduledChannel> {
+		self.channel_scheduler.lock().await.get_next_channel(payjoin_amount)
+	}
+
+	fn accept_payjoin_without_channel_opening(
+		&self, provisional_proposal: ProvisionalProposal,
+	) -> Result<String, Error> {
+		// fixme: instead of psbt.clone() we should sign the psbt,
+		// but BDK adds bip32 derivation paths to the psbt which is not supported by payjoin
+		let finalized_proposal =
+			provisional_proposal.finalize_proposal(|psbt| Ok(psbt.clone()), None)?;
+		Ok(finalized_proposal.psbt().to_string())
+	}
+
+	async fn handle_incoming_payjoin_request(
+		&self, request: Request<impl Body>,
+	) -> Result<String, Error> {
+		let (body, query, headers) = self.extract_request_parts(request).await?;
+		let provisional_proposal =
+			self.validate_payjoin_request(body.as_slice(), &query, headers.clone(), None)?;
+		let payjoin_amount = provisional_proposal.payjoin_amount();
+		if let Some(channel) = self.get_next_channel(payjoin_amount).await {
+			let mut proposal = provisional_proposal.clone();
+			let output_script = channel.output_script().expect("Output script should exist");
+			// substitute the output script with the channel output script
+			proposal.substitute_output_script(output_script);
+			let finalized_proposal =
+				proposal.clone().finalize_proposal(|psbt| Ok(self.sign_tx(&psbt)?), None)?;
+			let psbt = finalized_proposal.psbt();
+			let funding_tx = psbt.clone().extract_tx();
+			// tell the channel_scheduler that the funding tx has been created
+			self.channel_scheduler
+				.lock()
+				.await
+				.mark_as_funding_tx_created(channel.user_channel_id(), psbt.clone());
+			let temporary_channel_id =
+				channel.temporary_channel_id().expect("Temporary channel id should exist");
+			// tell the counterparty node that the funding tx has been created
+			let _ = self.receiver_handler.funding_transaction_generated(
+				&temporary_channel_id,
+				channel.node_id(),
+				funding_tx.clone(),
+			)?;
+			// wait for the counterparty node to return FundingSigned message
+			match tokio::time::timeout(tokio::time::Duration::from_secs(3), async move {
```

Comment:
The only way for me to know if the `FundingSigned` was received is by catching the tx in the `tx_broadcaster` https://github.com/lightningdevkit/ldk-node/pull/257/commits/d682093b62bbbc623c20b8fe210bec85dafcabd9#diff-b707fa4c317cfc8afdfeb972e8ae1b6fd4dc36361289cc53703af17d08dd6630R53 so I wait here a bit until the `tx_broadcaster` updates the `ChannelScheduler` state.. need a better way to handle this